client/error.py

```diff
@@ -240,6 +277,71 @@ def to_text(self) -> str:
         column = click.style(str(self.column), fg="yellow")
         return f"{path}:{line}:{column} {self.description}"
 
+    def to_sarif(self) -> Dict[str, Any]:
+        return {
+            "ruleId": "PYRE-ERROR-" + str(self.code)
+            if self.code is not None
+            else "PYRE-ERROR-MDL",
+            "level": "error",
+            "message": {"text": self.description},
+            "locations": [
+                {
+                    "physicalLocation": {
+                        "artifactLocation": {
+                            "uri": str(self.path) if self.path is not None else None,
+                        },
+                        "region": {
+                            "startLine": self.line,
+                            "startColumn": self.column + 1,
+                            "endLine": self.stop_line,
+                            "endColumn": self.stop_column,
+                        },
+                    },
+                },
+            ],
+        }
+
+    def get_sarif_rule(self) -> Dict[str, Any]:
+        return {
+            "id": "PYRE-ERROR-" + str(self.code)
+            if self.code is not None
+            else "PYRE-ERROR-MDL",
+            "name": "ModelVerificationError",
+            "shortDescription": {"text": "Model verification error"},
+            "helpUri": "https://www.pyre-check.org",
+            "help": {"text": "Model Verification error"},
+        }
+
+
+def errors_to_sarif(
+    errors: Union[
+        Sequence[Error], Sequence[LegacyError], Sequence[ModelVerificationError]
+    ]
+) -> Dict[str, Any]:
+
+    results: List[Dict[str, Any]] = []
+    rules: List[Dict[str, Any]] = []
+    for error in errors:
+        results.append(error.to_sarif())
+        rules.append(error.get_sarif_rule())
+
+    return {
+        "version": "2.1.0",
+        "$schema": "http://json.schemastore.org/sarif-2.1.0-rtm.4.json",
+        "runs": [
+            {
+                "tool": {
+                    "driver": {
+                        "name": "Pyre",
+                        "informationUri": "https://www.pyre-check.org",
+                        "rules": list({rule["id"]: rule for rule in rules}.values()),
```

Comment:
Is this trying to deduplicate rules by id? Would be nice to have a comment there.

Reply:
Yes! I will add a comment here.